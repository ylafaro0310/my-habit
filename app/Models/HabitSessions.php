<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use App\Models\BaseModel;

class HabitSessions extends BaseModel {
    public $validationRules = [
        'completed_at' => 'required|integer',
    ];

    public $tableName = 'habit_sessions';
}