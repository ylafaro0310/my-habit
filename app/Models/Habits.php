<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class Habits {
    static $validationRules = [
        'habit_name' => 'required|string|max255',
        'repeat_type' => 'required|string|in:day,week,day_of_week',
        'repeat_value' => 'required|integer',
        'started_at' => 'required|integer',
        'target_time' => 'nullable|integer',
        'time_of_day' => 'required|string|in:always,am,pm',
        'consecutive_days' => 'required|integer',
        'consecutive_weeks' => 'nullable|integer',
    ];

    static $tableName = 'habits';

    function __construct(){
        $this->pdo = DB::connection()->getPdo();
    }
    
    public function index(){
        $sth = $this->pdo->prepare("select * from ".self::$tableName);
        $sth->execute();
        $result = $sth->fetchAll();
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return $v===NULL?"NULL":"'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into ".self::$tableName." (${columns}) values (${values})");
        return $sth->execute();
    }

    public function update($id,$params){
        $setValues = array_map(function($key,$value){
            return $value === NULL ? "$key = NULL" : "$key = '$value'";
        },array_keys($params),array_values($params));
        $setValues = implode(', ',$setValues);
        $sth = $this->pdo->prepare("update ".self::$tableName." set $setValues where id = $id");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from ".self::$tableName." where id = ${id}");
        return $sth->execute();
    }
}