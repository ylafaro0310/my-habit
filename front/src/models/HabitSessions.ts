import { Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';

export default class HabitSessions extends Record<{
    habit_id: number;
    completed_at: Dayjs;
    working_hours: number; // 単位は「分」
}>({
    habit_id: 0,
    completed_at: dayjs(),
    working_hours: 0,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new HabitSessions(params);
  }
}