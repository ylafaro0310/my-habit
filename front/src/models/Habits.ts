import { List, Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';
import HabitRecords from './HabitRecords';

export class Habit extends Record<{
    id: number;
    habit_name: string;
    repeat_type: string;
    repeat_value: number;
    started_date: Dayjs;
    target_time: number; // 単位は「分」
    time_of_day: string;
    consecutive_days: number;
    consecutive_weeks: number | null;
}>({
    id: 0,
    habit_name: '',
    repeat_type: 'day_of_week',
    repeat_value: 127,
    started_date: dayjs(),
    target_time: 0,
    time_of_day: '',
    consecutive_days: 0,
    consecutive_weeks: null,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new Habit(params);
  }
  isCompleted(date: Dayjs,habit_records: HabitRecords): boolean {
    return habit_records.filterById(this.id).filterByCompletedAt(date = dayjs('2020-01-08')).getList().size > 0
  }
}

export default class Habits extends Record<{
  items: List<Habit>;
}>({
  items: List(),
}){
  static fromResponse(response: JSObject) {
    const params = { ...response };
    params.items = List(params.habits.map((item: JSObject) => Habit.fromResponse(item)));
    return new Habits(params);
  }
  getList(): List<Habit>{
    return this.get('items');
  }
}