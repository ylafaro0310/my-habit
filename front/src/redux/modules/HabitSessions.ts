import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import HabitSessions from '../../models/HabitSessions';

// Action Creator
const actionCreator = actionCreatorFactory('HabitSession');
export const HabitSessionsActions = {
  getHabitSessions: actionCreator<{ habitId: number; values?: object }>(
    'getHabitSessions',
  ),
  setHabitSessions: actionCreator<HabitSessions>('setHabitSessions'),
  addHabitSession: actionCreator<{ habitId: number; values: object }>(
    'addHabitSession',
  ),
  updateHabitSession: actionCreator<{
    habitId: number;
    habitSessionId: number;
    values: object;
  }>('updateHabitSession'),
  removeHabitSession: actionCreator<{
    habitId: number;
    habitSessionId: number;
  }>('removeHabitSession'),
  formInitialize: actionCreator<{ habitId: number; habitSessionId: number }>(
    'habitSessionFormInitialize',
  ),
};

// Reducers
export const HabitSessionsReducer = reducerWithInitialState(
  new HabitSessions(),
).case(HabitSessionsActions.setHabitSessions, (state, payload) => {
  return state.set('items', payload.getList());
});
