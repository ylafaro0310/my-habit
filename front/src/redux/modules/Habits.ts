import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import Habits from '../../models/Habits';

// Action Creator
const actionCreator = actionCreatorFactory('Habit');
export const HabitsActions = {
  getHabits: actionCreator<void>('getHabits'),
  setHabits: actionCreator<Habits>('setHabits'),
  addHabit: actionCreator<object>('addHabit'),
  updateHabit: actionCreator<{ habitId: number; values: object }>(
    'updateHabit',
  ),
  removeHabit: actionCreator<number>('removeHabit'),
  formInitialize: actionCreator<number>('habitFormInitialize'),
};

// Reducers
export const habitsReducer = reducerWithInitialState(new Habits()).case(
  HabitsActions.setHabits,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
