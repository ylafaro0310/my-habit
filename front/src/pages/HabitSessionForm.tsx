import React from 'react';
import { Dispatch } from 'redux';
import { Field, InjectedFormProps, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';

import { State } from '../redux/store';
import dayjs from '../lib/dayjs-ja';
import { HabitSessionsActions } from '../redux/modules/HabitSessions';
import {
  ButtonDanger,
  ButtonPrimary,
  CustomText,
  Form,
} from '../components/Form';
import SessionHeader from '../components/SessionHeader';
import Habits from '../models/Habits';

type HabitSessionFormProps = {
  habits: Habits;
  habitSession: object;
  dispatch: Dispatch;
} & InjectedFormProps<{}, HabitSessionFormProps> &
  RouteComponentProps<{ habitId: string; habitSessionId: string }>;

const initialValues = {
  id: 0,
  workingMinutes: 0,
  completedAt: dayjs().format('YYYY-MM-DD HH:mm:ss'),
};

export class HabitSessionForm extends React.Component<HabitSessionFormProps> {
  componentDidMount() {
    const { habitId, habitSessionId } = this.props.match.params;
    const { dispatch } = this.props;
    if (habitId && habitSessionId) {
      dispatch(
        HabitSessionsActions.formInitialize({
          habitId: Number(habitId),
          habitSessionId: Number(habitSessionId),
        }),
      );
    }
  }

  onClickRemove(
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    habitId: number,
    habitSessionId: number,
  ): void {
    event.preventDefault();
    const { dispatch } = this.props;
    const isYes = window.confirm(
      'この習慣に関する全てのデータが削除されます。この習慣を削除しますか？',
    );
    if (isYes) {
      dispatch(
        HabitSessionsActions.removeHabitSession({ habitId, habitSessionId }),
      );
    }
  }

  render() {
    const { habits, handleSubmit } = this.props;
    const { habitId, habitSessionId } = this.props.match.params;
    return (
      <>
        <SessionHeader
          backTo={'/habits/' + habitId + '/sessions/list'}
          habitName={
            habits
            ? habits.getList().find(elem => elem.id === Number(habitId))
              ?.habitName || ''
            : ''
          }
        />
        <Form onSubmit={handleSubmit}>
          <div>
            <Field
              component={CustomText}
              label='作業時間'
              name='workingMinutes'
            />
          </div>
          <div>
            <Field component={CustomText} label='完了日時' name='completedAt' />
          </div>
          <ButtonPrimary type='submit'>
            {habitSessionId ? '更新する' : '追加する'}
          </ButtonPrimary>
          {habitSessionId ? (
            <ButtonDanger
              onClick={e => {
                this.onClickRemove(e, Number(habitId), Number(habitSessionId));
              }}
              type='button'
            >
              セッションを削除する
            </ButtonDanger>
          ) : null}
        </Form>
      </>
    );
  }
}

export default connect((state: State) => ({
  initialValues: initialValues,
  habits: state.habits,
}))(
  reduxForm<{}, HabitSessionFormProps>({
    form: 'habitSession',
    onSubmit: (values, dispatch, props) => {
      const { habitId, habitSessionId } = props.match.params;
      const params = {
        habitId: Number(habitId),
        habitSessionId: Number(habitSessionId),
        values,
      };
      if (habitSessionId) {
        dispatch(HabitSessionsActions.updateHabitSession(params));
      } else {
        dispatch(HabitSessionsActions.addHabitSession(params));
      }
    },
  })(HabitSessionForm),
);
