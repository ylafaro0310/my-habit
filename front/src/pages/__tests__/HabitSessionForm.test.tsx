import React from 'react';
import { shallow } from 'enzyme';
import MockDate from 'mockdate';

import dayjs from '../../lib/dayjs-ja';
import { HabitSessionForm } from '../HabitSessionForm';
import Habits from '../../models/Habits';
import HabitSessions from '../../models/HabitSessions';

MockDate.set('2020-02-08');

const data = {
  habits: [
    { id: 1, habitName: '読書する', startedAt: dayjs() },
    { id: 2, habitName: '筋トレ', startedAt: dayjs() },
    { id: 3, habitName: '新しいCDを1枚聴く', startedAt: dayjs() },
  ],
  habitSessions: [
    { habitId: 2, workingMinutes: 5, completedAt: dayjs() },
    { habitId: 1, workingMinutes: 10, completedAt: dayjs('2020-01-08') },
  ],
};

const props = {
  habits: Habits.fromResponse(data),
  habitSessions: HabitSessions.fromResponse(data),
  match: {
    params: {
      id: 1,
    },
  },
  dispatch: jest.fn(),
};

describe('Page: HabitSessionForm', () => {
  it('Snapshot test', () => {
    const wrapper = shallow(<HabitSessionForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
