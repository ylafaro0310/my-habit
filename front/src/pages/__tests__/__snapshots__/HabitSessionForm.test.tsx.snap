// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: HabitSessionForm Snapshot test 1`] = `
<Fragment>
  <HabitHeader
    backTo="/habits/undefined/sessions/list"
    habitName=""
  />
  <Form>
    <div>
      <Field
        component={[Function]}
        label="作業時間"
        name="workingMinutes"
      />
    </div>
    <div>
      <Field
        component={[Function]}
        label="完了日時"
        name="completedAt"
      />
    </div>
    <ButtonPrimary
      type="submit"
    >
      追加する
    </ButtonPrimary>
  </Form>
</Fragment>
`;
